### ISSUE TITLE
Enable Java preview features

### ISSUE THREAD

**Issue by oskar2517** [jplag/JPlag#1303]
I am working with submissions that use [pattern matching in switch statements](https://docs.oracle.com/en/java/javase/17/language/pattern-matching-switch-expressions-and-statements.html#GUID-E69EEA63-E204-41B4-AA7F-D58B26A3B232), which is a language feature hidden behind the `--enable-preview` flag in Java 17. Is it possible to tell JPlag to enable preview features when parsing Java files? Currently, it fails with the following error message:
```
failed to parse 'MyClass.java' with reason: patterns in switch statements are a preview feature and are disabled by default.
  (use --enable-preview to enable patterns in switch statements)
        at de.jplag.ParsingException.wrappingExceptions(ParsingException.java:70)
        at de.jplag.java.JavacAdapter.parseFiles(JavacAdapter.java:59)
        at de.jplag.java.Parser.parse(Parser.java:25)
        at de.jplag.java.Language.parse(Language.java:47)
        at de.jplag.Submission.parse(Submission.java:249)
        at de.jplag.SubmissionSet.parseSubmissions(SubmissionSet.java:147)
        at de.jplag.SubmissionSet.parseAllSubmissions(SubmissionSet.java:103)
        at de.jplag.SubmissionSet.<init>(SubmissionSet.java:45)
        at de.jplag.SubmissionSetBuilder.buildSubmissionSet(SubmissionSetBuilder.java:78)
        at de.jplag.JPlag.run(JPlag.java:55)
        at de.jplag.cli.CLI.main(CLI.java:65)
```

**Comment by SirYwell** [comment]
It's currently not possible to pass arguments to the language module directly, and I think this isn't something that can be added easily. Additionally, the token extraction does not cover preview features, so while passing the option would turn the error into a warning, plagiarism detection might not work as expected in such cases. 

There are also breaking changes between the preview features in Java 17 and their finalized versions in Java 21, so supporting preview features likely causes additional work twice. 

You *can* run JPlag with Java 21, from a quick test that seems to work, but the problem of the preview feature having slight syntactical and semantical differences will persist (it might result in different compilation errors).

Alternatively, you can probably either fork or extend the existing Java language module to suite your needs, I'm not sure if the JPlag maintainers are interested in such changes.

(@tsaglam this would be a case where a lexer-based language module is benefitial - with the aforementioned limitations)

**Comment by oskar2517** [comment]
> It's currently not possible to pass arguments to the language module directly, and I think this isn't something that can be added easily. Additionally, the token extraction does not cover preview features, so while passing the option would turn the error into a warning, plagiarism detection might not work as expected in such cases. 

That means, plagiarism in switch expressions would likely not be detected?

**Comment by SirYwell** [comment]
> That means, plagiarism in switch expressions would likely not be detected?

All the code elements should be reached, so the code of each case will be handled just fine, but the pattern itself is not covered - whether this helps or hurts the detection is unclear.

**Comment by oskar2517** [comment]
I think this is fine for now. Thank you.

I have forked the repository and added the required flags to enable preview features in the parser.

**Comment by tsaglam** [comment]
We could probably consider supporting preview features with JPlag. I will take this into our internal discussions.

**Comment by tsaglam** [comment]
Closed by #1370.

**Comment by euberdeveloper** [comment]
I still have this error: 

```
2024-02-06-06:46:53_541 [main] [WARN] Submission - Failed to parse submission gr2 with error {}
de.jplag.ParsingException: failed to parse '/XXX/ConnectionHandler.java' with reason: patterns in switch statements are a preview feature and are disabled by default.
  (use --enable-preview to enable patterns in switch statements)
        at de.jplag.java.JavacAdapter.lambda$processErrors$1(JavacAdapter.java:81)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
        at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
        at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
        at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
        at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
        at de.jplag.java.JavacAdapter.processErrors(JavacAdapter.java:82)
        at de.jplag.java.JavacAdapter.parseFiles(JavacAdapter.java:57)
        at de.jplag.java.Parser.parse(Parser.java:25)
        at de.jplag.java.Language.parse(Language.java:47)
        at de.jplag.Submission.parse(Submission.java:249)
        at de.jplag.SubmissionSet.parseSubmissions(SubmissionSet.java:147)
        at de.jplag.SubmissionSet.parseAllSubmissions(SubmissionSet.java:103)
        at de.jplag.SubmissionSet.<init>(SubmissionSet.java:45)
        at de.jplag.SubmissionSetBuilder.buildSubmissionSet(SubmissionSetBuilder.java:78)
        at de.jplag.JPlag.run(JPlag.java:55)
        at de.jplag.cli.CLI.main(CLI.java:91)
```

I downloaded the latest jar in the Releases section

**Comment by koppor** [comment]
There is an error message when using Java 23. Does this relate to this usse?

```
2024-12-02-12:04:54_836 [WARN] Submission - Failed to parse submission repo079:
failed to parse '<null>'error: invalid source release 21 with --enable-preview
  (preview language features are only supported for release 23)
```

**Comment by tsaglam** [comment]
@koppor What is your exact Java version? What source code did you parse (or which preview feature)?

**Comment by koppor** [comment]
> @koppor What is your exact Java version? What source code did you parse (or which preview feature)?

I used JDK23. When switching to JDK21, it worked fine.

**Comment by SirYwell** [comment]
The `--enable-preview` flag only works when combined with the `--release X` flag, where `X` matches the version of the running JVM. JPlag currently hard-codes that value to 21, so it won't work on any other version.

In theory it's possible to set the version dynamically, but that defeats the purpose of the flag as it means there might be incompatibilities (i.e. Java 21 and 22 can deal with string templates, but Java 23 can't).

**Comment by koppor** [comment]
> The `--enable-preview` flag only works when combined with the `--release X` flag, where `X` matches the version of the running JVM. JPlag currently hard-codes that value to 21, so it won't work on any other version.

I did not specify any flags. Just executed jplag with following command line

    java -jar jplag.jar ../lecture/ -l java -bc ../lecture/example/ -s api/src

Thus, no `--enable-preview`. Just had JDK23 as `JAVA_HOME` - and in the `PATH`.

**Comment by misan** [comment]
Similar problem here with JDK22. It seems I have to downgrade to 21.

**Comment by misan** [comment]
Yes, I can confirm that going back to JDK21 from JDK22 made the error
```
[WARN] Submission - Failed to parse submission repo079:
failed to parse '<null>'error: invalid source release 21 with --enable-preview
  (preview language features are only supported for release 22)
```
go away and the same  jplag-5.1.0-jar-with-dependencies.jar file to work ok.

**Comment by SirYwell** [comment]
> I did not specify any flags. Just executed jplag with following command line
> 
> ```
> java -jar jplag.jar ../lecture/ -l java -bc ../lecture/example/ -s api/src
> ```
> 
> Thus, no `--enable-preview`. Just had JDK23 as `JAVA_HOME` - and in the `PATH`.

Sorry @koppor, I should have clarified that this is a problem inside of JPlag. JPlag itself uses these flags when passing the code to the javac tool here https://github.com/jplag/JPlag/blob/3fb0b8fd8d289a7f78bc21e5074d372eea164987/languages/java/src/main/java/de/jplag/java/JavacAdapter.java#L44-L45

**Comment by tsaglam** [comment]
I did not have the chance to fully read all comments here as I am currently out of office, but the issues might be resolved with #1854. The next release will contain this fix, which will most likely be released before the end of February. If your issues above are not addressed by #1854, we will take a look at them.

**Comment by tsaglam** [comment]
@misan @koppor, the issue should be fixed with the version we released yesterday: [JPlag v6.0.0](https://github.com/jplag/JPlag/releases/tag/v6.0.0). If you still encounter any problems, please let me know!

**Comment by misan** [comment]
@tsaglam I do not have a problem with v6.0.0 jar and JDK23. Thanks for the update!
